### ISSUE TITLE
[SelectableText] inconsistent selection background color between different character encoding

### ISSUE THREAD

**Issue by renntbenrennt** [flutter/flutter#68675]
## Steps to Reproduce
Select text when text containing English and Chinese, the selection color isn't cover the all the text while part of it is smaller if English is selected while another part of it is bigger if Chinese is selected.

**Expected results:** 
![image](https://user-images.githubusercontent.com/13610283/96661703-4ab3be00-137f-11eb-9bc7-5005f49e89ab.png)

**Actual results:** 
![image](https://user-images.githubusercontent.com/13610283/96661572-f1e42580-137e-11eb-822a-38c470027c5a.png)

**Comment by darshankawar** [comment]
@SeasonLeee 
Please provide `flutter doctor -v` and a complete minimal reproducible code sample that shows the issue.
Thanks.

**Comment by SeasonLeee** [comment]
No problem, here's the `flutter doctor -v`:
![image](https://user-images.githubusercontent.com/13610283/96711389-10264180-13d0-11eb-8d8e-daaa3670c786.png)

And I don't think this issue is sdk specific, it's been there for a while I just don't have the time to open a issue to talk about it, thanks

**Comment by darshankawar** [comment]
@SeasonLeee 
It's probably the way the selection toolbar works depending on the language and it's height. Since the English letters are a bit smaller than Chinese characters, it interprets and selects based on the height. As an example, Russian is same height as English, so the selection height would be same, but if you see the chinese characters, they are bigger and hence the difference.

**Comment by SeasonLeee** [comment]
Yeah, it could be, but I can't be sure the code to paint that background color, could you point me to the file that control the painting? thanks! @darshankawar

**Comment by darshankawar** [comment]
@SeasonLeee 
Check [this](https://github.com/flutter/flutter/tree/master/packages/flutter/lib/src/painting) if it helps.

**Comment by SeasonLeee** [comment]
@darshankawar oh, man, it's too much... 😅... I just tried to track down something related to SelectableText... but... ugh... I feel like I'm lost in the jungle...😅, thanks, though...

**Comment by jamesblasco** [comment]
I think this would be possible to edit by using the param `selectionHeightStyle : BoxHeightStyle.max` in `EditableText`;

https://github.com/flutter/flutter/blob/955e07461bd8a428310f381c4e4f1ca27653cdb1/packages/flutter/lib/src/widgets/editable_text.dart#L1161-L1164

Sadly this param is not accessible from SelectableText, TextField, TextFormField, CupertinoTextField.'

@justinmc, What would you think about adding this param to the [TextSelectionThemeData](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/text_selection_theme.dart)?

I also noticed that `selectionHeightStyle : BoxHeightStyle.max` should be the default one in Cupertino. Now all EditableText default to  `selectionHeightStyle : BoxHeightStyle.tight`

Apple Mail App: 

<img height="400" src="https://user-images.githubusercontent.com/19904063/96747694-e289ca00-13c8-11eb-880e-85392b89c201.jpeg" />

**Comment by SeasonLeee** [comment]
@jamesblasco Wow! words are useless when I want to use them to express my amazement on how quick you locate the cause was!!!! Man, you brilliant!!!

And yes, I agree with that we should expose the setting of `selectionHeightStyle ` or at least make it default to `BoxHeightStyle.max` instead of the original `BoxHeightStyle.tight`

Although, the reason why `tight` is the default maybe the fact that there's some problem with `max` when the text has change into new line if I understand correctly?

As stated in the codebase's comment:
![image](https://user-images.githubusercontent.com/13610283/96753449-77f58000-1402-11eb-934b-3afa3c4fb7f3.png)

**Comment by jamesblasco** [comment]
Thanks for sharing the full comments, reading more slowly the comment it looks like `includeLineSpacingMiddle` would be the way to go.  I would need to try more in advanced

**Comment by darshankawar** [comment]
Keeping this issue open for further analysis.

**Comment by justinmc** [comment]
That seems to make sense to me.  @darrenaustin what do you think about being able to configure selectionHeightStyle in TextSelectionThemeData?

Also CC @GaryQian in case there are any text height considerations.

**Comment by darrenaustin** [comment]
> That seems to make sense to me. @darrenaustin what do you think about being able to configure selectionHeightStyle in TextSelectionThemeData?

That sounds reasonable to me. If we do add it the `TextSelectionThemeData`, we should probably also include the 'selectionWidthStyle' as well to be consistent. As for the defaults for these, I am not familiar enough with the text system to have much of an opinion, but the default of `tight` leads to some awkward behavior here.

**Comment by jamesblasco** [comment]
Yes, I had also in mind to add `selectionWidthStyle`. Makes totally sense to add both at the same time.

I can do a PR with this change. @SeasonLeee would you be interested in give it a try and do the PR?

**Comment by SeasonLeee** [comment]
@jamesblasco 
😆oh my god!!! thanks for mentioning me and giving me this chance, and of course I would love to give it a shot!!!! god, I'm just so excited and this is so exhilarating!!!!

**Comment by SeasonLeee** [comment]
I would like to be assigned to this issue😁

**Comment by nt4f04uNd** [comment]
this also creates some small gaps/overlaps as seen on this screenshot

![flutter_selection_tearing](https://user-images.githubusercontent.com/39104740/112690427-cc5acb00-8e8c-11eb-8417-45b64d126d13.png)

**Comment by nt4f04uNd** [comment]
another note: overall selection height seems to be adjusted to the maximum height of a single character in the string

| just text | text with tall chars |
| - | - |
| ![just_text](https://user-images.githubusercontent.com/39104740/112758954-896a3600-8ff9-11eb-84ae-8e9a718147e1.png) | ![text_with_tall_chars](https://user-images.githubusercontent.com/39104740/112758961-8d965380-8ff9-11eb-892a-f1280a5db7e7.png) |

**Comment by LongXiangGuo** [comment]
Thanks very much, this issue had been fixed in flutter 2.5.3, test with the latest SelectableText, it works well.

**Comment by huanghui1998hhh** [comment]
Why not add `selectionHeightStyle` to `TextSelectionThemeData`? Or make `SelectionArea` have `selectionHeightStyle`.

**Comment by mafreud** [comment]
Is it okay to set ```selectionHeightStyle : BoxHeightStyle.includeLineSpacingMiddle``` as a default?
